Clustered particle point lists

re-smeis opened the pull request:
- LoadRexFileExample was modified to allow loading of multiple .rex files at once
- Implemented prefabs for clustered and non-clustered particleSystem-based pointListObjects
- Implemented threading for the creation of clustered particleSystem-based pointListObjects to reduce spikes in frame-times

re-smeis committed:
New branch, copied over from deprecated
added clustered particle system renderer for rex point list objects

re-smeis committed:
Implemented threading for clusteredParticlesPrefab

re-cheid commented on Assets/Example/LoadRexFileExample.unity:
```diff
@@ -363,5 +379,7 @@ MonoBehaviour:
   m_EditorClassIdentifier: 
-  rexFileName: oscar.rex
+  rexFileNames:
+  - D&S_Foyer_P3DMKII_Modellkoordinatensystem.rex
```
please change the scene again to a file which is part of this repository ;)

re-cheid commented on Assets/FreeFloatEditorCamera.cs:
```diff
@@ -0,0 +1,67 @@
+
+using UnityEngine;
+
+namespace RoboticEyes.Rex.ARSystem
```
this script can be placed in a new folder "Assets/Development" and the namespace can be removed

re-cheid commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,124 @@
+				Random.Range (min.z, max.z));
+		}
+
+		public override void SetRendererEnabled (bool enabled)
```
these methods need to be implemented

re-cheid commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListParticlesPrefab.cs:
```diff
@@ -0,0 +1,60 @@
+			return true;
+		}
+
+		public override void SetRendererEnabled (bool enabled)
```
these methods need to be implemented

re-cheid commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/ThreadedClustering.cs:
```diff
@@ -0,0 +1,84 @@
+using UnityEngine;
+using Random = System.Random;
+
+namespace RoboticEyes.Rex.RexFileReader.Examples
```
please add the copyright header here as well

re-smeis committed:
fixed formatting issues (tabs -> spaces)
added missing copyright headers
moved camera script to development folder and fixed namespace
implemented setLayer/setRenderer where missing
reverted default scene to use default model
added Rider specific gitignore entries
added missing meta files to VCS

re-ckart commented on Assets/Development/FreeFloatEditorCamera.cs:
```diff
@@ -0,0 +1,60 @@
+
+    [SerializeField] private Camera playerCamera;
+
+    [SerializeField] private bool lockYMovement = true;
```
Coding Convention:
```
[SerializeField] 
private float movementSpeed = 2f;
```

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,130 @@
+	public class PointListClusteredParticlesPrefab : RexPointListObject
+	{
+		[SerializeField] private int clusterCount = 16;
+		[SerializeField] private GameObject pointListParticlesPrefab;
```
SerliazeFiled into seperate line please

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,130 @@
+
+		private static List<ColoredPoint> CombinePointInfo (List<Vector3> pointPositions, List<Color> pointColors)
+		{
+			var lasPoints = pointColors.Select ((color, i) => new ColoredPoint {point = pointPositions[i], color = color}).ToList ();
```
variable name unclear. It refers to a specific file format (LAS), but the variable itself  (or the whole code in general) has no connection to said format.

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListParticlesPrefab.cs:
```diff
@@ -0,0 +1,67 @@
+{
+    public class PointListParticlesPrefab : RexPointListObject
+    {
+        [SerializeField] private bool pauseAfterInstantiation = true;
+        [SerializeField] private float particleSize = 0.1f;
+        [SerializeField] private ParticleSystem particlesSystem;
```
coding standard

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListParticlesPrefab.cs:
```diff
@@ -0,0 +1,67 @@
+                    startColor = pointColors[i],
+                    startSize = particleSize
+                };
+                particle.SetMeshIndex (0);
```
Why do we need to set this?

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListParticlesPrefab.cs:
```diff
@@ -0,0 +1,67 @@
+            if (enabled)
+            {
+                //necessary for particles to become visible
+                GetComponent<ParticleSystem> ().SetParticles (particles);
```
We already have the ParticleSystem as a variable, No need for GetComponent

re-ckart reviewed (changes requested):
![image](https://user-images.githubusercontent.com/36508191/115229507-6c65e600-a113-11eb-88a7-44030e15d834.png)
Deactivate Distance Sorting

re-smeis committed:
Adjusted formatting
LoadRexFileExample.cs now supports loading assets on Android
Fixed unnecessary casts
Simplified random centroid selection
Cluster count is now automatically calculated from maxAllowedParticles and averageParticlesPerCluster
Large point clouds are now automatically thinned out, with option for dynamic density in prefab (default: off)
Parallelized expensive operations where possible

re-ckart commented on Assets/Development/FreeFloatEditorCamera.cs:
```diff
@@ -12,18 +12,19 @@
         Destroy (this);
 #else
-        playerCamera.transform.position = new Vector3 (0f, 1f, 0f);
+        // playerCamera.transform.position = new Vector3 (0f, 1f, 0f);
```
could delete the #else path here

re-ckart commented on Assets/Example/LoadRexFileExample.cs:
```diff
@@ -37,53 +39,80 @@ private void Start()
 
-        if (meshesToWrite != null && meshesToWrite.Length > 0)
-        {
+        if (meshesToWrite!=null && meshesToWrite.Length > 0) {
```
Styling wrong. Spaces between operators missing. And brackets are not in new line.
This is also the case in other files!
Should be
``` C#
if (meshesToWrite != null && meshesToWrite.Length > 0)
{
```

re-ckart replied:
Line Below. I think WriteAllBytes would not work on Android. Maybe write only in editor or use persistentDatePath when **not** in editor

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,172 @@
+			);
+
+			return true;
+		}
```
Weird spaces in the whole file

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListParticlesPrefab.cs:
```diff
@@ -0,0 +1,127 @@
+        public void SetDensity(float densityFactor)
+        {
+            //prevent irrelevant updates
+            if(Mathf.Abs(densityFactor - currentDensityFactor) < 0.05f) return;
```
Coding style, should be
```
if(Mathf.Abs(densityFactor - currentDensityFactor) < 0.05f) 
{
  return;
}
```

re-ckart replied:
should the factor difference (0.05f) be set-able?

re-smeis replied:
Will add reasonably range-capped attribute for fine-tuning.

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/ThreadedClustering.cs:
```diff
@@ -0,0 +1,94 @@
+            return minIdx;
+        }
+
+        public struct ColoredPoint
```
Please move struct to the top of the code

re-cheid replied:
also I'm not sure whether using a `class` instead of `struct` might be better here.
structs are value types and pushed on the stack, classes are reference types on the heap.
Not sure though which is faster here / less memory

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/ThreadedClustering.cs:
```diff
@@ -0,0 +1,94 @@
+                clusteredPoints = new List<ColoredPoint>[kVal];
+                var concurrentClusteredPoints = InitializeConcurrentClusterPointsList (kVal);
+
+                Parallel.ForEach (pointPositions, pointPosition => {
+                    concurrentClusteredPoints[GetIndexOfClosestCentroid (pointPosition.point, centroids)].Add (pointPosition);
+                });
```
Could be made simpler:
``` C#
Parallel.ForEach (pointPositions, pointPosition => concurrentClusteredPoints[GetIndexOfClosestCentroid (pointPosition.point, centroids)].Add (pointPosition));
```

re-cheid replied:
does not matter here IMHO, would either result in a long line or a broken line without parentheses, both weird

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/ThreadedClustering.cs:
```diff
@@ -0,0 +1,94 @@
+
+        public override void WorkerFunction()
+        {
+            try {
```
bracket into new line

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,172 @@
+
+			StartClusteringCoroutine (coloredPoints, startingCentroids, (success, clusteredPointPositions) =>
+				{
+					if (!success) return;
```
``` C#
if( ...)
{
...
}
```

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,172 @@
+				yield return null;
+			}
+
+			childPointListParticles = GetComponentsInChildren<PointListParticlesPrefab> ();
```
Be careful, this would only return active Children. If you also want inactive one use `GetComponentsInChildren(true)`

re-ckart commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,172 @@
+		[SerializeField] 
+		private int averageParticlesPerCluster = (int)60e3;
+		[SerializeField] 
+		private GameObject pointListParticlesPrefab;
```
if you specify the Class instead of GameObject, you dont need to use GetComponent() when calling Instantiate later on.

re-cheid commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListClusteredParticlesPrefab.cs:
```diff
@@ -0,0 +1,172 @@
+		{
+			var sw = new Stopwatch ();
+			sw.Start ();
+			print ("starting clustering...");
```
do you mean `Debug.Log`?

re-smeis replied:
`public static void print(object message) => Debug.Log(message);`

It's just a built in wrapper in MonoBehaviour for debug.log but can change it for consistency

re-cheid replied:
yes please ;)

re-cheid commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/PointListParticlesPrefab.cs:
```diff
@@ -0,0 +1,127 @@
+
+        public override void SetRendererEnabled (bool enabled)
+        {
+            gameObject.SetActive (enabled);
```
you might try to enable/disable the MeshRenderer directly via `particlesSystem.shape.meshRenderer`, this might make the set particles below obsolete.
Should also speed up the process as gameobject de- and activation always performs some GC tasks

re-smeis replied:
> particlesSystem.shape.meshRenderer

good to know! will do

re-smeis replied:
`particlesSystem.shape.meshRenderer` appears to refer to the emission shape. I did find though that `particlesSystem.GetComponent<Renderer> ();` (so the renderer component of the particle system itself) does access the correct renderer, now it works =)

re-cheid commented on Assets/package/Runtime/Prefabs/StandardPrefabClasses/ThreadedClustering.cs:
```diff
@@ -0,0 +1,94 @@
+            }
+        }
+
+        private static ConcurrentBag<ColoredPoint>[] InitializeConcurrentClusterPointsList (int kVal)
```
Didn't know about `ConcurrentBag<T>`. I learned something today 😆

re-smeis committed:
Formatting changes, renamed some vars
Added more serialized fields for fine-tuning through inspector
Improved SetRendererEnabled function
Fixed potential crash under Android in LoadRexFileExample

re-smeis committed:
Formatting
